packages/service-library/src/servicelib/docker_utils.py

```diff
@@ -281,3 +287,70 @@ async def pull_image(
                 f"pulling {image_short_name}: {pull_progress}...",
                 logging.DEBUG,
             )
+
+
+def get_lifespan_remote_docker_client(
+    docker_api_proxy_settings_property_name: str,
+) -> Callable[[FastAPI], AsyncContextManager[None]]:
+    """Ensures `setup` and `teardown` for the remote docker client.
+
+    Arguments:
+        docker_api_proxy_settings_property_name -- if the name is `PROP_NAME`
+        then it should be accessible as `app.state.settings.PROP_NAME`
+
+    Returns:
+        docker client lifespan manager
+    """
+
+    @asynccontextmanager
+    async def _(app: FastAPI) -> AsyncIterator[None]:
+        settings: DockerApiProxysettings = getattr(
+            app.state.settings, docker_api_proxy_settings_property_name
+        )
+
+        session: ClientSession | None = None
+        if settings.DOCKER_API_PROXY_USER and settings.DOCKER_API_PROXY_PASSWORD:
+            session = ClientSession(
+                auth=aiohttp.BasicAuth(
+                    login=settings.DOCKER_API_PROXY_USER,
+                    password=settings.DOCKER_API_PROXY_PASSWORD.get_secret_value(),
+                )
+            )
+
+        async with AsyncExitStack() as exit_stack:
+            if settings.DOCKER_API_PROXY_USER and settings.DOCKER_API_PROXY_PASSWORD:
+                await exit_stack.enter_async_context(
+                    ClientSession(
+                        auth=aiohttp.BasicAuth(
+                            login=settings.DOCKER_API_PROXY_USER,
+                            password=settings.DOCKER_API_PROXY_PASSWORD.get_secret_value(),
+                        )
+                    )
+                )
+
+            client = await exit_stack.enter_async_context(
+                aiodocker.Docker(url=settings.base_url, session=session)
+            )
+
+            app.state.remote_docker_client = client
+
+            await wait_till_docker_api_proxy_is_responsive(app)
+
+            yield
```

Comment:
MINOR:  in the past we had nasty bugs for wrongly indenting this. I would add a small comment saying that this is inside exit_stack